wrappers/wrapper.rb

```diff
@@ -835,17 +835,285 @@ def check_distances(vrp, unfeasible)
+      #         end
+      #         nil # returns nil because the objects are modified and this function is going to be called automatically
+      #       end
+      #
+      # (If some modes are not necessary they can be merged -- e.g. `when :rewind, :patch_result` and have `nil`)
+      # :patch_result is called for interim solutions and for the last solution before the :rewind is called.
+
+      # TODO: We can simplify service timewindows if they are not necessary -- e.g., all service TWs are "larger" than
+      # the vehicle TWs. (this modification needs to be rewinded incase we are in dicho or max_split)
+
+      # TODO: infeasibility detection can be done with the simplification interface
+      # (especially the part that is done after matrix)
+
+      # Warning: The order might be important if the simplifications are interdependent.
+      # The simplifications will be called in the following order and their corresponding rewind
+      # and result patching operations will be called in the opposite order. This can be changed
+      # if necessary.
+      [
+        :simplify_vehicle_duration,
+        :simplify_vehicle_pause,
+      ].freeze
+    end
+
     def simplify_constraints(vrp)
-      if vrp[:vehicles] && !vrp[:vehicles].empty?
-        vrp[:vehicles].each{ |vehicle|
-          if (vehicle[:force_start] || vehicle[:shift_preference] == 'force_start') && vehicle[:duration] && vehicle[:timewindow]
-            vehicle[:timewindow][:end] = vehicle[:timewindow][:start] + vehicle[:duration]
-            vehicle[:duration] = nil
-          end
+      simplifications.each{ |simplification|
+        self.send(simplification, vrp, nil, mode: :simplify)
+      }
+
+      vrp
+    end
+
+    def patch_simplified_constraints_in_result(result, vrp)
+      return result unless result.is_a?(Hash)
+
+      simplifications.reverse_each{ |simplification|
+        self.send(simplification, vrp, result, mode: :patch_result)
+      }
+
+      result
+    end
+
+    def patch_and_rewind_simplified_constraints(vrp, result)
+      # first patch the results (before the constraint are rewinded)
+      patch_simplified_constraints_in_result(result, vrp) if result.is_a?(Hash)
+
+      # then rewind the simplifications
+      simplifications.reverse_each{ |simplification|
+        self.send(simplification, vrp, nil, mode: :rewind)
+      }
+
+      vrp
+    end
+
+    def simplify_vehicle_duration(vrp, _result = nil, options = { mode: :simplify })
+      # Simplify vehicle durations using timewindows if there is force_start
+      case options[:mode]
+      when :simplify
+        vrp.vehicles&.each{ |vehicle|
+          next unless (vehicle.force_start || vehicle.shift_preference == 'force_start') && vehicle.duration && vehicle.timewindow
+
+          # Warning: this can make some services infeasible because vehicle cannot work after tw.start + duration
+          vehicle.timewindow.end = vehicle.timewindow.start + vehicle.duration
```

Comment:
This is not implemented in the current PR and I just added a warning and did the syntax refacto. Otherwise, the code is the same. So I don't know.

I assumed we don't let vehicles with `v.duration > v.tw.end - v.tw.start` pass the validation check in the first place because it is discordant. But I don't see such a verification.

I think letting  `v.duration > v.tw.end - v.tw.start` pass is dangerous because it ignores possible data errors. I believe this case is not possible on mt-web due to verifications done on import and vehicle page, so we can have the same verification without risk. What do you guys think, do I include in this PR or not?

Reply:
`v.duration > v.tw.end - v.tw.start` is not really discordant but redundant. The "correction" is obvious from our point of view. I see no reason to return a validation error in this case.

Reply:
It is redundant AND discordant. A vehicle cannot work upto `v.duration` if `v.duration > v.tw.end - v.tw.start` because the duration limit doesn't agree with the time window.

If we silently ignore the duration then we are setting `v.duration = nil` which is dangerous because we don't know what the main intention of the user was -- there are three possibilities.

1- they might have sent the vehicle.duration unintentionally, in that case the user should remove the vehicle duration, not us
2- it might be an error in the v.duration (i.e., v.duration should have been less than the TW) then the user can see and correct it
3- it might be an error in the v.TW (i.e., v.TW should have been at least as large as v.duration) then the user can see and correct it

I don't see why optim-api should accept a VRP which has more chance to be incorrect than correct, silently modify the VRP and return a result which will potentially be discarded -- especially while the upstream refusing this exact situation.

Reply:
In any case, this is not something introduced in this dev and out of scope of the current PR, I suggest you create a ticket @braktar and we will discuss how to handle this situation there.

Reply:
vehicle.duration expresses that the route muse be contained between 0 and this duration. Any value within this interval is correct. If end - start is smaller than duration, just reduce the domain. It doesn't make the domain empty.

Imagine the following use case.
A driver must work 7 hours each day. With an timewindow 8:00 - 18:00
Any minute not used on a day is reported on the following days.
I might happen that on friday the driver has a credit of work duration above the ten hours. It is not incorrect from the point of view of the user. And this situation may not appear all along the study phase but in the production one.

Reply:
I understand the use case. Still, this doesn't mean that optim-api should accept such VRPs. It is not optim-api's responsibility to keep track of such info, but letting the user know about possible discrepancy/désaccord between their parameters is.

The user doesn't need optim-api to keep track of the these extra credits, they can do that on their side (even inside json that they sent us with an unused key like `vehicle[:credit]`). The added benefit is that they can apply whatever logic that would make sense for their use case. 

The issue is we cannot use dependent parameter logic because of poor performance of grape, so the grape-swagger can't see this verification so neither the documentation nor the auto generated classes can prevent the user from generating an invalid VRP on their side :/